simulator/func_sim/operation.h

```diff
@@ -113,6 +113,15 @@ class Operation
     void set_v_imm( uint32 value) { v_imm = value; }
     auto get_delayed_slots() const { return delayed_slots; }
 
+    Addr get_decoded_target() const
+    {
+        if ( this->is_branch() )
+            return this->PC + 4 + Addr{ narrow_cast<uint64>(this->v_imm) } * 4;
```

Comment:
That has to be adjusted for the case of delayed branches, please see the new code in `alu.h`

Reply:
???

Reply:
As I understand, function ```get_decoded_target``` must return ``` instr->new_PC = instr->PC + 4 * (1 + instr->get_delayed_slots());``` in case when branch is not taken.

But we don't know if it is taken or not at decode stage.
Furthermore, I think that function ```get_decoded_target``` must consider only jumps as we accelerate only them at decode stage.

Reply:
OK, I see. I applies only to taken branches.

Reply:
@YanLo 
No, I was too fast. In delayed branch mode, delayed slots **are always** considered, regardless of taken or not taken.

Additionally, that string contains one more bug: immediate is not sign extended, i.e. branches which point backwards should always have the target misprediction.
I'll fix in the next patch along with few other changes.

Actually, that was the only line not covered by any of the tests...

Reply:
Can you please explain me why should we leave considering branch in this function?
Because we use ```get_decoded_target``` only in case of jump.
As I see, delayed slots are applied only for branches.

Reply:
> Because we use get_decoded_target only in case of jump.

No. Branch may get target misprediction as well.

> As I see, delayed slots are applied only for branches.

Where have you read that?

Reply:
>No. Branch may get target misprediction as well.

Yes. But if it is predicted as not taken and actually not taken, flush, caused by target misprediction will be a wrong action.

Reply:
> Yes. But if it is predicted as not taken and actually not taken, flush, caused by target misprediction will be a wrong action.

What if predicted as taken but the target is wrong (e.g. due to aliasing with another branch)?

Reply:
If we flush in this case we still don't know what instruction to fetch, because we don't know at **decode** stage is it **actually** taken or not.

Reply:
Yes. If the branch is actually not taken, we will flush at condition resolution. If branch is actually taken, we fetched wrong code anyways.

Reply:
(Actually, I thought you have that pipelines studied somewhere on paper even before you had written a line of code).

Reply:
Yes, and I am trying to explain why I think that we consider misprediction at **decode** stage only in case of **jump**